PermissionsScrapper/PermissionsScrapper/PermsDescription.cs

```diff
@@ -0,0 +1,141 @@
+                    {
+                        _scopesDescriptionsJson = scopesDescriptionsJson;
+
+                        // TODO: Call GitHub API with json file to upload to devx-content-repo
+                    }
+                }
+                else
+                {
+                    log.LogInformation($"Failed to get authentication at: {DateTime.Now}");
+                }
+            }
+            catch (Exception ex)
+            {
+                log.LogInformation($"Exception occurred: {ex.InnerException.Message ?? ex.Message}\r\n Time of occurrence: {DateTime.Now}");
+            }
+        }
+
+        /// <summary>
+        /// Gets authentication to a protected Web API.
+        /// </summary>
+        /// <param name="config">The application configuration settings.</param>
+        /// <returns>An authentiction result, if successful.</returns>
+        private static AuthenticationResult GetAuthentication(ApplicationConfig config)
+        {
+            IConfidentialClientApplication app;
+            app = ConfidentialClientApplicationBuilder.Create(config.ClientId)
+                  .WithClientSecret(config.ClientSecret)
+                  .WithAuthority(new Uri(config.Authority))
+                  .Build();
+
+            string[] scopes = new string[] { $"{config.ApiUrl}.default" };
+
+            try
+            {
+                return app.AcquireTokenForClient(scopes)
+               .ExecuteAsync().GetAwaiter().GetResult();
+            }
+            catch (Exception ex)
+            {
+                throw ex;
+            }
+        }
+
+        /// <summary>
+        /// Retrieves permissions and their descriptions from a Service Principal
+        /// </summary>
+        /// <param name="config">The application configuration settings.</param>
+        /// <param name="result">The JSON response of the permissions and their descriptions retrieved from the Service Prinicpal.</param>
+        /// <returns></returns>
+        private static string GetScopesDescriptionsJson(ApplicationConfig config, AuthenticationResult result)
+        {
+            try
+            {
+                var httpClient = new HttpClient();
+                var apiCaller = new ProtectedApiCallHelper(httpClient);
+                var spJson = apiCaller
+                    .CallWebApiAsync($"{config.ApiUrl}{config.ApiVersions[0]}/serviceprincipals?$filter=appId eq '{config.ServicePrincipalId}'", result.AccessToken)
+                    .GetAwaiter().GetResult(); // fetch for v1.0 ; no business case for fetching for beta yet
+
+                var cleanedSpJson = CleanJsonData(spJson, config);
+
+                // Retrieve the respective scopes
+                var spValue = JsonConvert.DeserializeObject<JObject>(cleanedSpJson).Value<JArray>("value");
+
+                var scopesDescriptions = new Dictionary<string, List<Dictionary<string, object>>>();
```

Comment:
This is a very nested data type. Do you mind adding what an example structure looks like? What is a key for top level dictionary and a key for second level dictionary, etc..

Reply:
Added helper comments in code. I hope this helps. The structure (simplified) looks like this:
```
{
    "value": [
        {            
            "appRoles": [
                {
                    "allowedMemberTypes": [
                        "Application"
                    ],
                    "description": "Allows a Teams app to read, install, upgrade, and uninstall itself to any user, without a signed-in user.",
                    "displayName": "Allow the app to manage itself for all users",
                    "id": "908de74d-f8b2-4d6b-a9ed-2a17b3b78179",
                    "isEnabled": true,
                    "origin": "Application",
                    "value": "TeamsAppInstallation.ReadWriteSelfForUser.All"
                }
            ],            
            "oauth2PermissionScopes": [
                {
                    "adminConsentDescription": "Allows the app to read, install, upgrade, and uninstall Teams apps installed for the signed-in user. Does not give the ability to read application-specific settings.",
                    "adminConsentDisplayName": "Manage user's installed Teams apps",
                    "id": "093f8818-d05f-49b8-95bc-9d2a73e9a43c",
                    "isEnabled": true,
                    "type": "Admin",
                    "userConsentDescription": "Allows the app to read, install, upgrade, and uninstall Teams apps installed for you. Does not give the ability to read application-specific settings.",
                    "userConsentDisplayName": "Manage your installed Teams apps",
                    "value": "TeamsAppInstallation.ReadWriteForUser"
                }
            ]
        }
    ]
}
```